File: src/main/java/dhasday/adventofcode/common/AStarSearch.java
package dhasday.adventofcode.common;

import java.util.HashMap;
import java.util.LinkedList;
import java.util.List;
import java.util.Map;
import java.util.Set;
import java.util.function.BiFunction;
import java.util.function.Function;

import javafx.util.Pair;

public class AStarSearch<T> {

    public List<T> findShortestPath(T start,
                                    T end,
                                    BiFunction<T, T, Integer> heuristicCostEstimate,
                                    Function<T, Set<Pair<T, Integer>>> findAdjacentNodes) {
        Map<T, AStarNode> closedSet = new HashMap<>();
        Map<T, AStarNode> openSet = new HashMap<>();

        AStarNode startNode = new AStarNode(start, null, 0, heuristicCostEstimate.apply(start, end));
        openSet.put(start, startNode);

        while (!openSet.isEmpty()) {
            AStarNode currentNode = getNodeWithLowestFScore(openSet);
            if (currentNode.value.equals(end)) {
                return reconstructPathToNode(currentNode);
            }

            openSet.remove(currentNode.value);
            closedSet.put(currentNode.value, currentNode);

            Set<Pair<T, Integer>> adjacentValues = findAdjacentNodes.apply(currentNode.value);
            for (Pair<T, Integer> adjacentValueDistancePair : adjacentValues) {
                T adjacentValue = adjacentValueDistancePair.getKey();
                Integer adjacentDistance = adjacentValueDistancePair.getValue();

                if (closedSet.containsKey(adjacentValue)) {
                    continue;
                }

                AStarNode adjacentNode = openSet.get(adjacentValue);
                int possibleGScore = currentNode.gScore + adjacentDistance;

                if (adjacentNode == null || adjacentNode.gScore < possibleGScore) {
                    adjacentNode = new AStarNode(
                            adjacentValue,
                            currentNode,
                            possibleGScore,
                            heuristicCostEstimate.apply(adjacentValue, end)
                    );
                }

                openSet.put(adjacentValue, adjacentNode);
            }
        }

        return null;
    }

    private AStarNode getNodeWithLowestFScore(Map<T, AStarNode> openSet) {
        AStarNode lowestNode = null;

        for (AStarNode node : openSet.values()) {
            if (lowestNode == null || node.fScore < lowestNode.fScore) {
                lowestNode = node;
            }
        }

        return lowestNode;
    }

    private List<T> reconstructPathToNode(AStarNode node) {
        List<T> path = new LinkedList<>();
        path.add(0, node.value);

        if (node.previousNode != null) {
            path.addAll(0, reconstructPathToNode(node.previousNode));
        }

        return path;
    }

    private class AStarNode {
        private T value;
        private AStarNode previousNode;

        private int gScore;
        private int hScore;
        private int fScore;

        AStarNode(T value, AStarNode previousNode, int gScore, int hScore) {
            this.value = value;
            this.previousNode = previousNode;

            this.gScore = gScore;
            this.hScore = hScore;
            this.fScore = gScore + hScore;
        }
    }
}

File: src/main/python/common/a_star_search.py
class AStarSearch(object):

    class AStarNode(object):
        def __init__(self, value, previous_node, g_score, h_score):
            self.value = value
            self.previous_node = previous_node

            self.g_score = g_score
            self.h_score = h_score
            self.f_score = g_score + h_score

    def find_shortest_path(self, start, end, heuristic_cost_estimate, find_adjacent_nodes):
        """
        Returns the shortest path from start to end using A* Search Algorithm

        :param start: The initial node
        :param end: The target node
        :param heuristic_cost_estimate: A function that takes 2 nodes and calculates the expected
                                        cost to get from the first to the second
        :param find_adjacent_nodes: A function that takes 1 node and returns pairs of all nodes
                                    that are directly accessible from that node and their distance
                                    ( node, distance )
        :return: The shortest path from start to end
        """
        closed_set = dict()
        open_set = dict()

        start_node = self.AStarNode(start, None, 0, heuristic_cost_estimate(start, end))
        open_set[start] = start_node

        while open_set:
            current_node = self._get_node_with_lowest_f_score(open_set)
            if current_node.value == end:
                return self._reconstruct_path_to_node(current_node)

            del open_set[current_node.value]
            closed_set[current_node.value] = current_node

            adjacent_values = find_adjacent_nodes(current_node.value)
            for (adjacent_value, adjacent_distance) in adjacent_values:
                if adjacent_value in closed_set:
                    continue

                adjacent_node = open_set.get(adjacent_value)
                possible_g_score = current_node.g_score + adjacent_distance

                if adjacent_node is None or adjacent_node.g_score < possible_g_score:
                    adjacent_node = self.AStarNode(
                        adjacent_value,
                        current_node,
                        possible_g_score,
                        heuristic_cost_estimate(adjacent_value, end)
                    )

                open_set[adjacent_value] = adjacent_node

        # We've explored everything and there's no path from start to end
        return None

    def _get_node_with_lowest_f_score(self, open_set):
        lowest_node = None

        for node in open_set.values():
            if lowest_node is None or node.f_score < lowest_node.f_score:
                lowest_node = node

        return lowest_node

    def _reconstruct_path_to_node(self, node):
        path = list()

        cur_node = node
        while cur_node is not None:
            path.insert(0, cur_node.value)
            cur_node = cur_node.previous_node

        return path
